File: kmp/shared/src/commonMain/kotlin/app/cadence/shared/domain/PartOfDay.kt
package app.cadence.shared.domain

import kotlinx.datetime.LocalTime

/**
 * The half of «Воскресенье, утро» that isn't the weekday, computed here (not on the screen)
 * as a rule about time. Boundaries are the ones Russian actually uses, not even quarters:
 * «ночь» to five, «утро» to noon, «день» to six, «вечер» to midnight.
 */
enum class PartOfDay(
    val ru: String,
) {
    NIGHT("ночь"),
    MORNING("утро"),
    AFTERNOON("день"),
    EVENING("вечер"),
}

private const val MORNING_FROM = 5
private const val AFTERNOON_FROM = 12
private const val EVENING_FROM = 18

fun partOfDay(time: LocalTime): PartOfDay =
    when (time.hour) {
        in MORNING_FROM until AFTERNOON_FROM -> PartOfDay.MORNING
        in AFTERNOON_FROM until EVENING_FROM -> PartOfDay.AFTERNOON
        in EVENING_FROM..LAST_HOUR -> PartOfDay.EVENING
        else -> PartOfDay.NIGHT
    }

private const val LAST_HOUR = 23

File: api/internal/protocol/partofday.go
package protocol

import "github.com/SimonOsipov/cadence-app/api/internal/platform/civil"

// PartOfDay is the half of «Воскресенье, утро» that is not the weekday: a rule about the
// clock, computed here rather than on the screen so both surfaces say the same word.
type PartOfDay string

const (
	Night     PartOfDay = "night"
	Morning   PartOfDay = "morning"
	Afternoon PartOfDay = "afternoon"
	Evening   PartOfDay = "evening"
)

// The boundaries Russian actually uses, and not even quarters: «ночь» to five, «утро» to
// noon, «день» to six, «вечер» to midnight. Ported from the KMP rather than rounded.
const (
	morningFrom   = 5
	afternoonFrom = 12
	eveningFrom   = 18
)

// PartOfDayAt is the word for a time of day. The slot is the patient's own — the caller
// computes it in their zone, the way every day in this feature is computed.
func PartOfDayAt(at civil.Slot) PartOfDay {
	switch {
	case at.Hour >= morningFrom && at.Hour < afternoonFrom:
		return Morning
	case at.Hour >= afternoonFrom && at.Hour < eveningFrom:
		return Afternoon
	case at.Hour >= eveningFrom:
		return Evening
	default:
		return Night
	}
}
